src/Build/BackEnd/Components/Scheduler/Scheduler.cs

```diff
@@ -2280,8 +2271,37 @@ private void AssignGlobalRequestId(BuildRequest request)
                 }
             }
 
-            request.GlobalRequestId = _nextGlobalRequestId;
-            _nextGlobalRequestId++;
+            // No matching request found, assign a new ID
+            request.GlobalRequestId = _nextGlobalRequestId++;
+        }
+
+        /// <summary>
+        /// Computes a hash code for a collection of targets, ignoring order and case.
+        /// </summary>
+        private int ComputeTargetsHash(List<string> targets)
+        {
+            int hash = 17; // Start with a non-zero seed
+            foreach (string target in targets)
+            {
+                hash = hash * 31 + StringComparer.OrdinalIgnoreCase.GetHashCode(target);
+            }
+
+            return hash;
+        }
+
+        /// <summary>
+        /// Determines if two target collections contain the same targets, ignoring order and case.
+        /// </summary>
+        private bool TargetsMatch(List<string> firstTargetsList, List<string> secondTargetsList)
+        {
+            if (firstTargetsList.Count != secondTargetsList.Count)
+            {
+                return false;
+            }
+
+            HashSet<string> set = new HashSet<string>(firstTargetsList, StringComparer.OrdinalIgnoreCase);
+
+            return set.SetEquals(secondTargetsList);
```

Comment:
There's a boxed enumerator for the call to `SetEquals`. That might not be worth worrying about, but it's worth profiling to see if it causes too many allocations in this path.